src/middlewared/middlewared/plugins/container/migrate.py

```diff
@@ -1,15 +1,143 @@
 from middlewared.api.current import (
     ContainerMigrateArgs, ContainerMigrateResult,
 )
-from middlewared.service import CallError, job, Service
+from middlewared.service import CallError, job, private, Service
 from middlewared.plugins.pool_.utils import UpdateImplArgs
 from middlewared.plugins.zfs.mount_unmount_impl import MountArgs
+from middlewared.plugins.zfs.rename_promote_clone_impl import RenameArgs
+
+from .utils import container_dataset
+
+
+class VirtGlobalModel(sa.Model):
+    """Legacy virt_global table model for migration purposes."""
+    __tablename__ = 'virt_global'
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    pool = sa.Column(sa.String(120), nullable=True)
+    storage_pools = sa.Column(sa.Text(), nullable=True)
+    bridge = sa.Column(sa.String(120), nullable=True)
+    v4_network = sa.Column(sa.String(120), nullable=True)
+    v6_network = sa.Column(sa.String(120), nullable=True)
 
 
 class ContainerService(Service):
+
+    @private
+    async def migrate_devices(self, job, devices, container_instance):
+        container_name = container_instance["name"]
+        nic_choices = await self.middleware.call("container.device.nic_attach_choices")
+        usb_choices = await self.middleware.call("container.device.usb_choices")
+        gpu_choices = await self.middleware.call("container.device.gpu_choices")
+        for device_name, device_data in devices.items():
+            dtype = None
+            try:
+                device_payload = None
+                dtype = device_data.get("type")
+                if dtype == "disk":
+                    src = device_data.get("source", "")
+                    if src.startswith("/mnt") is False:
+                        job.logs_fd.write((
+                            f"Skipping migrating {device_name!r} disk device for {container_name!r} because "
+                            f"source does not start with /mnt/ (is {src!r} instead)\n"
+                        ).encode())
+                        continue
+
+                    device_payload = {
+                        "dtype": "FILESYSTEM",
+                        "source": src,
+                        "target": device_data["path"],
+                    }
+                elif dtype == "nic":
+                    if device_data.get("parent") not in nic_choices:
+                        job.logs_fd.write((
+                            f"Skipping migrating {device_name!r} NIC device for {container_name!r} because "
+                            f"{device_data.get('parent')!r} is not a valid NIC\n"
+                        ).encode())
+                        continue
+
+                    device_payload = {
+                        "dtype": "NIC",
+                        "nic_attach": device_data["parent"],
+                        "type": "VIRTIO",
+                    }
+                elif dtype == "usb":
+                    if (bus_num := device_data.get("busnum")) and (devnum := device_data.get("devnum")):
```

Comment:
Should we skip this validation? There's no big deal in having a container with a nonexistent device IIRC. However, USB devices can be detached, so this code will remove the devices that are not available at the time the upgrade is run even if the USB device was just removed temporarily.